SignalServiceKit/src/Loki/Protocol/Friend Requests/FriendRequestProtocolTests.swift

```diff
@@ -23,46 +25,619 @@ class FriendRequestProtocolTests : XCTestCase {
         TSAccountManager.sharedInstance().didRegister()
     }
 
-    func testMultiDeviceFriendRequestAcceptance() {
-        // When Alice accepts Bob's friend request, she should accept all outstanding friend requests with Bob's
-        // linked devices and try to establish sessions with the subset of Bob's devices that haven't sent a friend request.
-        func getDevice() -> DeviceLink.Device? {
-            guard let publicKey = Data.getSecureRandomData(ofSize: 64) else { return nil }
-            let hexEncodedPublicKey = "05" + publicKey.toHexString()
-            guard let signature = Data.getSecureRandomData(ofSize: 64) else { return nil }
-            return DeviceLink.Device(hexEncodedPublicKey: hexEncodedPublicKey, signature: signature)
+    // MARK: - Helpers
+
+    func isFriendRequestStatus(_ values: [LKFriendRequestStatus], for hexEncodedPublicKey: String, transaction: YapDatabaseReadWriteTransaction) -> Bool {
+        let status = storage.getFriendRequestStatus(for: hexEncodedPublicKey, transaction: transaction)
+        return values.contains(status)
+    }
+
+    func isFriendRequestStatus(_ value: LKFriendRequestStatus, for hexEncodedPublicKey: String, transaction: YapDatabaseReadWriteTransaction) -> Bool {
+        return isFriendRequestStatus([value], for: hexEncodedPublicKey, transaction: transaction)
+    }
+
+    func generateHexEncodedPublicKey() -> String {
+        return Curve25519.generateKeyPair().hexEncodedPublicKey
+    }
+
+    func getDevice(for hexEncodedPublicKey: String) -> DeviceLink.Device? {
+        guard let signature = Data.getSecureRandomData(ofSize: 64) else { return nil }
+        return DeviceLink.Device(hexEncodedPublicKey: hexEncodedPublicKey, signature: signature)
+    }
+
+    func createContactThread(for hexEncodedPublicKey: String) -> TSContactThread {
+        var result: TSContactThread!
+        storage.dbReadWriteConnection.readWrite { transaction in
+            result = TSContactThread.getOrCreateThread(withContactId: hexEncodedPublicKey, transaction: transaction)
+        }
+        return result
+    }
+
+    func createGroupThread(groupType: GroupType) -> TSGroupThread? {
+        let stringId = Randomness.generateRandomBytes(kGroupIdLength)!.toHexString()
+        let groupId: Data!
```

Comment:
Also, this doesn't need to be an implicitly unwrapped optional. The compiler realizes that you're assigning it in every case below that doesn't end execution before the value is used.